### ISSUE TITLE
cq_id 0 is out of range error when test finishes

### ISSUE THREAD

**Issue by s-jovic** [tenstorrent/tt-metal#8965]
When running `test_falcon_mlp.py` component test, it very often (almost always) errors out with the following error after the devices are closed and pytest prints out passed:

```
                 Always | FATAL    | cq_id 0 is out of range
terminate called after throwing an instance of 'std::runtime_error'
  what():  TT_FATAL @ ../tt_metal/impl/device/device.cpp:1673: cq_id < sw_command_queues_.size()
info:
cq_id 0 is out of range
```

Sometimes just segmentation error is shown, but no description. Llama team noticed a similar error in their code (one of the last messages on the model channel by @kevinmiTT11 ).

I was able to reproduce it in a unit-test (if you don't get it on the first run, try a couple of more times, it almost always errors out on the first time for me). The machines I tried are all with N300 (bgd-lab-07, bgd-lab-05 and one of the Austin cloud machines). I tried on main as of May 30 (https://github.com/tenstorrent/tt-metal/commit/e75540b8ffc224fa4ce0c0aee4ba2cb21040f715)/

```
import ttnn
import torch
from models.demos.falcon7b.tt.model_config import get_model_config

def test_deallocate_bug(device):
    seq_len = 1024
    model_config = get_model_config("BFLOAT16-DRAM", seq_len)

    tt_padding = torch.zeros((1, 1, seq_len, 64)).bfloat16().float()  # 4608 - 4544 = 64
    tt_padding = ttnn.from_torch(
        tt_padding,
        ttnn.bfloat16,
        device=device,
        layout=ttnn.TILE_LAYOUT,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )
    model_config["MLP_PREFILL_PADDING_TENSORS"] = tt_padding

    # uncomment to avoid segfault
    # model_config.pop("MLP_PREFILL_PADDING_TENSORS")

    print("Tensors initialized")
```

The error goes away if we explicitly pop the entry from the dictionary (you can uncomment second to last line in the test to try it out). I think the issue seems to be that we put tensor into a dictionary, and when Python garbage collection tries to deallocate the dictionary, we get this error. The only specific thing to Falcon7b here is the model_config object; replacing the object with empty dictionary makes the error go away. Using Bfp8 instead of 16 still results in an error.

**Comment by s-jovic** [comment]
cc @skhorasganiTT

**Comment by xanderchin** [comment]
@tt-aho @abhullar-tt I recall there was another deallocate issue that you were both commenting on before. Is this similar? 

Otherwise, the issue assignment may be misplaced.

**Comment by abhullar-tt** [comment]
It seems like the sw command queues get destructed before the tensor is getting deallocated in the segfault case. Segfault is seen here because device is closed so sw command queue/allocator gets destructed but when the tensor attempts to deallocate the buffer, the buffer will try to access the destructed objects within device.

We ran into this issue before because device doesn't have any references to the buffers that get allocated on it. Wondering if  we should change this and then on device::close deallocate all buffers? We would need to "invalidate" accesses to this buffer/tensor once this is done. Thoughts @tt-aho?

**Comment by tt-aho** [comment]
> It seems like the sw command queues get destructed before the tensor is getting deallocated in the segfault case. Segfault is seen here because device is closed so sw command queue/allocator gets destructed but when the tensor attempts to deallocate the buffer, the buffer will try to access the destructed objects within device.
> 
> We ran into this issue before because device doesn't have any references to the buffers that get allocated on it. Wondering if we should change this and then on device::close deallocate all buffers? We would need to "invalidate" accesses to this buffer/tensor once this is done. Thoughts @tt-aho?

I think this should work. It's weird that python is not destructing the tensors in the map before device close since they should have went out of scope before we close device

**Comment by bbradelTT** [comment]
> > It seems like the sw command queues get destructed before the tensor is getting deallocated in the segfault case. Segfault is seen here because device is closed so sw command queue/allocator gets destructed but when the tensor attempts to deallocate the buffer, the buffer will try to access the destructed objects within device.
> > We ran into this issue before because device doesn't have any references to the buffers that get allocated on it. Wondering if we should change this and then on device::close deallocate all buffers? We would need to "invalidate" accesses to this buffer/tensor once this is done. Thoughts @tt-aho?
> 
> I think this should work. It's weird that python is not destructing the tensors in the dict before device close since they should have went out of scope before we close device

I think Python is like Java, it doesn't free memory until it wants to, and you can't guarantee when objects get destructed.

**Comment by tt-aho** [comment]
maybe we should try to force garbage collection in our device fixture before we close device.

```
import gc
gc.collect()
```

**Comment by bbradelTT** [comment]
> maybe we should try to force garbage collection in our device fixture before we close device.
> 
> ```
> import gc
> gc.collect()
> ```

That's worth a try. 

Hopefully there won't be an issue with some objects being kept for re-use (a workaround someone used was to create large objects: https://dinispeixoto.com/til/force-python-to-release-memory/). Otherwise, to avoid the problem, you'd probably need to keep a list of buffers to deallocate on close.

**Comment by tt-aho** [comment]
I think we should go with deallocating the list of buffers as the permanent solution so that it also prevents user error, not just python intricacies. Can try and see if forcing gc works as a stopgap until the deallocation is implemented

**Comment by aliuTT** [comment]
Don't we already have `this->deallocate_buffers();` in the device close?

**Comment by tt-aho** [comment]
That deallocates the blocks from the allocator, but buffers still holds/accesses ptrs to invalid objects

**Comment by aliuTT** [comment]
I'll pick up this thread of work. Going to add gc collect before device close from python side. This is blocking my other changes for DevicePool anyway. I'll try to get a clean CI run, but I was seeing many tests hanging with gc collection enabled 🌚

**Comment by tt-rkim** [comment]
@Shubhamsaboo @ttssokorac 
This is the original bug report, and the team is working on updating status in here

Other issue which affects front-page examples in README: #9066

**Comment by aliuTT** [comment]
@s-jovic can you provide a branch and test cmd that is giving you this segfault locally? I will try it out on my machine, along with a fix.

**Comment by aliuTT** [comment]
Fixed on main here `e62ab754f`. I've tried the test @s-jovic had in the original issue, and it was passing with this change cherry-picked. Closing issue now, re-open if anyone has further issues with this one. There are still some segfaults in other issues, might be unrelated.
